### DISCUSSION TITLE
Use severity to determine default rule weight?

### DISCUSSION THREAD

**Discussion by bsara** [detekt/detekt#4895] [Ideas]
Currently (v1.20.0), as far as I've been able to tell, the only way that I can get detekt to report as passing (using the gradle plugin) is to manually set the weight to zero of each rule that I don't want to be seen as an error (i.e. so that warnings and infos don't cause failures).

I think that I would make sense to have the severity of a rule determine the default weight of the rule, no? e.g. all rules that don't have a severity of error would have a default weight of zero. If one wants they could always override the default weight with the already provided way of setting weights.

On top of that, it would be nice if the console reporters and the html/text reports showed the severity so that when an error is present, one can distinguish between errors, warnings, and info's in the reports.

Sorry if there is a way to do this already, I can't seem to find it if there is an existing solution.

**Comment by cortinico** [comment]
You're right that there is no "easy" way to do this today, beyond what you suggested. 
We had a https://github.com/detekt/detekt/blob/main/detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/SeverityLevel.kt which we don't really use effectively, and we don't let the user customize it that much.

> I think that I would make sense to have the severity of a rule determine the default weight of the rule, no? e.g. all rules that don't have a severity of error would have a default weight of zero. If one wants they could always override the default weight with the already provided way of setting weights.

I think it will create more confusion at this stage, as we'll have to define arbitrary weight per severity which would force users to make calculations. Having everything weighted as 1 instead allows for further customization based on the codebase you're on.

**Reply by bsara** [reply]
> we'll have to define arbitrary weight per severity

I'm not proposing "arbitrary" default weights. I'm just proposing that only rules with a severity of "error" cause a failure. The best way I could see doing this with the correct implementation would be to make all non-error severities have a weight off zero by default.

I was actually a bit surprised when the task was seen as failed even though I had no error level severity rules violated. I initially thought it was a bug. I think that this is the intuitive expectation of users. For instance, my code will still compile and run even though I have warnings or suggestions of changes from my ide. Only when I have errors in my code will the action be considered to have failed (unless I've explicitly specified to fail on warnings as well). I would expect the same from a linter/static code analysis tool by default. Plus, I've never seen any other linter, by default, report as failed when only warnings or info suggestions are reported.

**Reply by bsara** [reply]
I think that, at the very least, it would be helpful to report the severity in the console and html reports. It be even more helpful if we could group violations by severity in those reports.

**Reply by bsara** [reply]
A config option could also be provided to allow the user to set the default weight of severities. Something like this perhaps:

```yaml
build:
  severityDefaultWeights:
    error: 42
    warning: 1
    info: 0
```

**Reply by bsara** [reply]
Another option might be something like a `failOnSeverities` option where one could list the severities that would be seen as causing a task to be seen as failed.

As you can probably tell, my main point is that we should be able to have task seen as failed based on severity not just manually set weights.

**Reply by cortinico** [reply]
> I'm just proposing that only rules with a severity of "error" cause a failure

I think we're missing a point here: there is no concept of "severity error" at the moment in Detekt. All the rules are treated equals. As I mentioned, there was a plan to implement `SeverityLevel` for every rule, but this is not implemented at the moment.

**Reply by bsara** [reply]
After reading through #3274, I see what you're saying.

Is there a plan to implement things according to these two comments in that issue?
- https://github.com/detekt/detekt/issues/3274#issuecomment-743775275
- https://github.com/detekt/detekt/issues/3274#issuecomment-747129088

It doesn't seem very clear to me what the long-term decision on the issue was there.

**Reply by cortinico** [reply]
I think we never came to a decision and this slowly got stale. I don't think we have a plan for this at the moment. Happy to re-evaluate if more users are requesting this specific issue

**Reply by bsara** [reply]
awesome! thanks

**Comment by schalkms** [comment]
I agree with @cortinico.

> I think it will create more confusion at this stage, as we'll have to define arbitrary weight per severity which would force users to make calculations. Having everything weighted as 1 instead allows for further customization based on the codebase you're on.

However, you could adjust the weight in the `detekt.yml` configuration via script based on the defined `SeverityLevel` in the source code of the corresponding rule.

Note that, we also had a discussion in #3274 for that topic. There was not much feedback from the community about changing this option, though.

**Reply by bsara** [reply]
See my responses to to @cortinico.
